R/method.R

```diff
@@ -1,145 +1,163 @@
+#'   * A base type specified either with its constructor (`logical`, `integer`,
+#'     `double` etc) or its name (`"logical"`, `"integer"`, "`double`" etc).
+#'   * A base union type specified by its name: `"numeric"`, `"atomic"`, or
+#'     `"vector"`.
+#'
+#'   For R7 generics that use multiple dispatch, this can be a list of any of
+#'   the above types.
+#'
+#'   For S3 generics, this must be an R7 class.
 #' @param value A function that implements the generic specification for the
-#'   given `signature`. The arguments must be compatible with the generic.
-#' @importFrom utils getS3method
+#'   given `signature`.
 #' @export
 #' @examples
 #' # Create a generic
 #' bizarro <- new_generic("bizarro", dispatch_args = "x")
 #' # Register some methods
-#' method(bizarro, "numeric") <- function(x, ...) rev(x)
-#' method(bizarro, s3_class("factor")) <- function(x, ...) {
-#'   levels(x) <- rev(levels(x))
-#'   x
-#' }
-#' method(bizarro, s3_class("data.frame")) <- function(x, ...) {
+#' method(bizarro, "numeric") <- function(x) rev(x)
+#' method(bizarro, s3_class("data.frame")) <- function(x) {
 #'   x[] <- lapply(x, bizarro)
 #'   rev(x)
 #' }
 #'
 #' # Using a generic calls the methods automatically
-#' bizarro(1)
-#'
-#' # But it can be useful to explicitly retrieve a method in order to
-#' # inspect its implementation
-#' method(bizarro, "double")
-#' method(bizarro, s3_class("factor"))
-method <- function(generic, signature) {
-  # TODO: check that signature doesn't contain any unions
-
-  signature <- as_signature(signature)
-  method_impl(generic, signature, ignore = NULL)
+#' bizarro(head(mtcars))
+`method<-` <- function(generic, signature, value) {
+  register_method(generic, signature, value)
+  invisible(generic)
 }
 
-methods <- function(generic) {
-  get_all_methods(generic@methods, character())
-}
+register_method <- function(generic, signature, method, package = packageName(parent.frame())) {
+  signature <- as_signature(signature)
+  generic <- as_generic(generic)
 
-get_all_methods <- function(x, signature) {
-  if(!is.environment(x)) {
-    return(x)
+  if (is_external_generic(generic)) {
+    if (!is.null(package)) {
+      # method registration within package, so add to lazy registry
+      external_methods_add(package, generic, signature, method)
+    } else {
+      # otherwise find the generic and register
+      generic <- getFromNamespace(generic$name, asNamespace(generic$package))
+      register_method(generic, signature, method)
+    }
+  } else if (is_s3_generic(generic)) {
+    generic_name <- attr(generic, "name")
+
+    if (length(signature) != 1 || class_type(signature[[1]]) != "r7") {
+      msg <- sprintf(
+        "When registering methods for S3 generic %s(), signature be a single R7 class",
+        generic_name
+      )
+      stop(msg, call. = FALSE)
+    }
+    class <- signature[[1]]@name
+    registerS3method(generic_name, class, method, envir = parent.frame())
```

Comment:
I am not sure if the use of `parent.frame()` here will always be right or not? Should this function have an `envir` argument so we can always pass an environment through as needed?